File: src/sttp/sttp-mcp.Tests/Services/SyncCoordinatorServiceTests.cs
using Microsoft.Extensions.Logging.Abstractions;
using Shouldly;
using SttpMcp.Application.Services;
using SttpMcp.Domain.Contracts;
using SttpMcp.Domain.Models;
using SttpMcp.Storage;

namespace SttpMcp.Tests.Services;

public sealed class SyncCoordinatorServiceTests
{
    [Fact]
    public async Task Pull_Should_Not_Resurface_RemoteRows_As_LocalChanges()
    {
        var ct = TestContext.Current.CancellationToken;
        var store = new InMemoryNodeStore();
        var cursorUpdatedAt = DateTime.Parse(
            "2026-03-05T06:41:00Z",
            null,
            System.Globalization.DateTimeStyles.RoundtripKind);

        var source = new StubChangeSource(new ChangeQueryResult
        {
            Nodes =
            [
                BuildNode("sync-session", "remote", "sync-a", "2026-03-05T06:41:00Z")
            ],
            NextCursor = new SyncCursor
            {
                UpdatedAt = cursorUpdatedAt,
                SyncKey = "sync-a"
            },
            HasMore = false
        });

        var coordinator = new SyncCoordinatorService(
            store,
            source,
            NullLogger<SyncCoordinatorService>.Instance);

        var result = await coordinator.PullAsync(new SyncPullRequest
        {
            SessionId = "sync-session",
            ConnectorId = "cloud-primary",
            PageSize = 50,
            MaxBatches = 1
        }, ct);

        result.Checkpoint.ShouldNotBeNull();
        result.Checkpoint!.Cursor.ShouldNotBeNull();

        var changes = await store.QueryChangesSinceAsync(
            "sync-session",
            result.Checkpoint.Cursor,
            50,
            ct);

        changes.Nodes.Count.ShouldBe(0);
        changes.HasMore.ShouldBeFalse();
    }

    [Fact]
    public async Task Pull_Should_Page_Changes_And_Advance_Checkpoint_Without_Owning_Policy()
    {
        var ct = TestContext.Current.CancellationToken;
        var store = new InMemoryNodeStore();
        var source = new StubChangeSource(new ChangeQueryResult
        {
            Nodes =
            [
                BuildNode("sync-session", "apply", "sync-a", "2026-03-05T06:31:00Z"),
                BuildNode("sync-session", "skip", "sync-b", "2026-03-05T06:32:00Z")
            ],
            NextCursor = new SyncCursor
            {
                UpdatedAt = DateTime.Parse("2026-03-05T06:32:00Z", null, System.Globalization.DateTimeStyles.RoundtripKind),
                SyncKey = "sync-b"
            },
            HasMore = false
        });

        var coordinator = new SyncCoordinatorService(
            store,
            source,
            NullLogger<SyncCoordinatorService>.Instance,
            new RejectSkipPolicy());

        var result = await coordinator.PullAsync(new SyncPullRequest
        {
            SessionId = "sync-session",
            ConnectorId = "cloud-primary",
            PageSize = 50,
            MaxBatches = 1
        }, ct);

        result.Fetched.ShouldBe(2);
        result.Created.ShouldBe(1);
        result.Filtered.ShouldBe(1);
        result.HasMore.ShouldBeFalse();
        result.Checkpoint.ShouldNotBeNull();
        result.Checkpoint!.Cursor.ShouldNotBeNull();
        result.Checkpoint.Cursor.SyncKey.ShouldBe("sync-b");

        var nodes = await store.QueryNodesAsync(new NodeQuery
        {
            SessionId = "sync-session",
            Limit = 10
        }, ct);

        nodes.Count.ShouldBe(1);
        nodes[0].Raw.ShouldBe("apply");
    }

    private sealed class StubChangeSource(ChangeQueryResult page) : ISyncChangeSource
    {
        private ChangeQueryResult? _page = page;

        public Task<ChangeQueryResult> ReadChangesAsync(
            string sessionId,
            string connectorId,
            SyncCursor? cursor,
            int limit,
            CancellationToken ct = default)
        {
            var page = _page ?? new ChangeQueryResult();
            _page = null;
            return Task.FromResult(page);
        }
    }

    private sealed class RejectSkipPolicy : ISyncCoordinatorPolicy
    {
        public bool ShouldAcceptNode(SttpNode node) => node.Raw != "skip";
    }

    private static SttpNode BuildNode(string sessionId, string raw, string syncKey, string updatedAt) => new()
    {
        Raw = raw,
        SessionId = sessionId,
        Tier = "raw",
        Timestamp = DateTime.Parse("2026-03-05T06:30:00Z", null, System.Globalization.DateTimeStyles.RoundtripKind),
        CompressionDepth = 1,
        ParentNodeId = null,
        SyncKey = syncKey,
        UpdatedAt = DateTime.Parse(updatedAt, null, System.Globalization.DateTimeStyles.RoundtripKind),
        SourceMetadata = null,
        UserAvec = new AvecState
        {
            Stability = 0.85f,
            Friction = 0.25f,
            Logic = 0.80f,
            Autonomy = 0.70f
        },
        ModelAvec = new AvecState
        {
            Stability = 0.91f,
            Friction = 0.21f,
            Logic = 0.90f,
            Autonomy = 0.80f
        },
        CompressionAvec = AvecState.Zero,
        Rho = 0.96f,
        Kappa = 0.94f,
        Psi = 2.6f
    };
}

File: src/sttp/sttp-core-rs/tests/sync_coordinator_service_tests.rs
use std::collections::VecDeque;
use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use tokio::sync::Mutex;

use sttp_core_rs::application::services::SyncCoordinatorService;
use sttp_core_rs::domain::contracts::{NodeStore, SyncChangeSource, SyncCoordinatorPolicy};
use sttp_core_rs::domain::models::{
    AvecState, ChangeQueryResult, SttpNode, SyncCursor, SyncPullRequest,
};
use sttp_core_rs::storage::InMemoryNodeStore;

struct StubChangeSource {
    pages: Mutex<VecDeque<ChangeQueryResult>>,
}

impl StubChangeSource {
    fn new(pages: Vec<ChangeQueryResult>) -> Self {
        Self {
            pages: Mutex::new(VecDeque::from(pages)),
        }
    }
}

#[async_trait]
impl SyncChangeSource for StubChangeSource {
    async fn read_changes_async(
        &self,
        _session_id: &str,
        _connector_id: &str,
        _cursor: Option<SyncCursor>,
        _limit: usize,
    ) -> Result<ChangeQueryResult> {
        Ok(self.pages.lock().await.pop_front().unwrap_or_default())
    }
}

struct RejectSkipPolicy;

impl SyncCoordinatorPolicy for RejectSkipPolicy {
    fn should_accept_node(&self, node: &SttpNode) -> bool {
        node.raw != "skip"
    }
}

fn build_test_node(session_id: &str, raw: &str, sync_key: &str, updated_at: &str) -> SttpNode {
    SttpNode {
        raw: raw.to_string(),
        session_id: session_id.to_string(),
        tier: "raw".to_string(),
        timestamp: DateTime::parse_from_rfc3339("2026-03-05T06:30:00Z")
            .expect("timestamp should parse")
            .with_timezone(&Utc),
        compression_depth: 1,
        parent_node_id: None,
        sync_key: sync_key.to_string(),
        updated_at: DateTime::parse_from_rfc3339(updated_at)
            .expect("timestamp should parse")
            .with_timezone(&Utc),
        source_metadata: None,
        context_summary: None,
        embedding: None,
        embedding_model: None,
        embedding_dimensions: None,
        embedded_at: None,
        user_avec: AvecState {
            stability: 0.85,
            friction: 0.25,
            logic: 0.80,
            autonomy: 0.70,
        },
        model_avec: AvecState {
            stability: 0.91,
            friction: 0.21,
            logic: 0.90,
            autonomy: 0.80,
        },
        compression_avec: Some(AvecState::zero()),
        rho: 0.96,
        kappa: 0.94,
        psi: 2.6,
    }
}

#[tokio::test(flavor = "current_thread")]
async fn pull_does_not_resurface_remote_rows_as_local_changes() {
    let store = Arc::new(InMemoryNodeStore::new());
    let source = Arc::new(StubChangeSource::new(vec![ChangeQueryResult {
        nodes: vec![build_test_node(
            "sync-session",
            "remote",
            "sync-a",
            "2026-03-05T06:41:00Z",
        )],
        next_cursor: Some(SyncCursor {
            updated_at: DateTime::parse_from_rfc3339("2026-03-05T06:41:00Z")
                .expect("timestamp should parse")
                .with_timezone(&Utc),
            sync_key: "sync-a".to_string(),
        }),
        has_more: false,
    }]));
    let coordinator = SyncCoordinatorService::new(store.clone(), source);

    let result = coordinator
        .pull_async(SyncPullRequest {
            session_id: "sync-session".to_string(),
            connector_id: "cloud-primary".to_string(),
            page_size: 50,
            max_batches: Some(1),
        })
        .await
        .expect("pull should succeed");

    let changes = store
        .query_changes_since_async(
            "sync-session",
            result
                .checkpoint
                .as_ref()
                .and_then(|checkpoint| checkpoint.cursor.clone()),
            50,
        )
        .await
        .expect("change query should succeed");

    assert!(changes.nodes.is_empty());
    assert!(!changes.has_more);
}

#[tokio::test(flavor = "current_thread")]
async fn coordinator_pages_changes_and_advances_checkpoint_without_owning_policy() {
    let store = Arc::new(InMemoryNodeStore::new());
    let source = Arc::new(StubChangeSource::new(vec![ChangeQueryResult {
        nodes: vec![
            build_test_node("sync-session", "apply", "sync-a", "2026-03-05T06:31:00Z"),
            build_test_node("sync-session", "skip", "sync-b", "2026-03-05T06:32:00Z"),
        ],
        next_cursor: Some(SyncCursor {
            updated_at: DateTime::parse_from_rfc3339("2026-03-05T06:32:00Z")
                .expect("timestamp should parse")
                .with_timezone(&Utc),
            sync_key: "sync-b".to_string(),
        }),
        has_more: false,
    }]));
    let coordinator =
        SyncCoordinatorService::with_policy(store.clone(), source, Arc::new(RejectSkipPolicy));

    let result = coordinator
        .pull_async(SyncPullRequest {
            session_id: "sync-session".to_string(),
            connector_id: "cloud-primary".to_string(),
            page_size: 50,
            max_batches: Some(1),
        })
        .await
        .expect("pull should succeed");

    assert_eq!(result.fetched, 2);
    assert_eq!(result.created, 1);
    assert_eq!(result.filtered, 1);
    assert!(!result.has_more);
    assert_eq!(
        result
            .checkpoint
            .as_ref()
            .and_then(|checkpoint| checkpoint.cursor.as_ref())
            .map(|cursor| cursor.sync_key.as_str()),
        Some("sync-b")
    );

    let stored = store
        .query_nodes_async(sttp_core_rs::domain::models::NodeQuery {
            limit: 10,
            session_id: Some("sync-session".to_string()),
            from_utc: None,
            to_utc: None,
            tiers: None,
        })
        .await
        .expect("query should succeed");
    assert_eq!(stored.len(), 1);
    assert_eq!(stored[0].raw, "apply");
}
